### PR TITLE
DNN : Add Exp layer

### PR DESCRIPTION
**Pull request by SamFC10**
Based on Caffe's [Exponential Layer](https://caffe.berkeleyvision.org/tutorial/layers/exp.html).

resolves: #19475 #18088 
merge with extra: opencv/opencv_extra#853

<cut/>

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2020.3.0:16.04
build_image:Custom Win=openvino-msvs2019-2021.2.0
build_image:Custom Mac=openvino-2021.2.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*

Xbuildworker:Custom=linux-4,linux-6
Xbuild_image:Custom=ubuntu-cuda:18.04
```

### CODE REVIEW

File: modules/dnn/src/layers/elementwise_layers.cpp
Comment (alalek) [inline modules/dnn/src/layers/elementwise_layers.cpp:1441]: We don't really need this code branch.

```
pow(base, exponent) = exp(ln(base) * exponent)  // base > 0
= exp(ln_base * (scale * INPUT + shift))
= exp((ln_base * scale) * INPUT + (ln_base * shift))
= exp(scale_normalized * INPUT + shift_normalized)
```

So add new "normalized" fields :
```
float ln_base = base == -1 ? 1 : ln(base);
float scale_normalized = scale * ln_base;
float shift_normalized = shift * ln_base;
```

and use `exp()` code only.

Comment (SamFC10) [reply modules/dnn/src/layers/elementwise_layers.cpp]: Done. Thanks for pointing it out.

File: modules/dnn/src/layers/elementwise_layers.cpp
Comment (alalek) [inline modules/dnn/src/layers/elementwise_layers.cpp:1412]: Please use CV_Check*() macros (instead of CV_Assert) to emit better error messages:

```
CV_Check(base, base == -1.f || base > 0.f, "Unsupported 'base' value");
```

**Review by alalek — COMMENTED** [review]
Thank you for update!

File: modules/dnn/src/layers/elementwise_layers.cpp
Comment (SamFC10) [reply modules/dnn/src/layers/elementwise_layers.cpp]: Fixed.

**Review by l-bat — APPROVED** [review]

**Comment by SamFC10** [comment]
@alalek Should I open a new PR to backport this layer to 3.4?

**Comment by alalek** [comment]
@SamFC10 This "backport" label is usually for maintainers.

Could you please squash commits into one before merge?

**Comment by SamFC10** [comment]
>This "backport" label is usually for maintainers.

Got it. Thanks for the info!

**Pull request merged**
